controllers/machinehealthcheck_controller_test.go

```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+			},
+		},
+	}
+	return node
+}
+
+func setNodeUnhealthy(node *corev1.Node) {
+	node.Status.Conditions[0] = corev1.NodeCondition{
+		Type:               corev1.NodeReady,
+		Status:             corev1.ConditionUnknown,
+		LastTransitionTime: metav1.NewTime(time.Now().Add(-10 * time.Minute)),
+	}
+}
+
+func newInfraMachine(machine *clusterv1.Machine) (*unstructured.Unstructured, string) {
+	providerID := fmt.Sprintf("test:////%v", uuid.NewUUID())
+	return &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "infrastructure.cluster.x-k8s.io/v1alpha3",
+			"kind":       "InfrastructureMachine",
+			"metadata": map[string]interface{}{
+				"generateName": "test-mhc-machine-infra-",
+				"namespace":    machine.Namespace,
+			},
+			"spec": map[string]interface{}{
+				"providerID": providerID,
+			},
+		},
+	}, providerID
+}
+
+func fakeNodesMachines(
+	g *WithT,
+	n int,
+	healthy bool,
+	assignNodeRef bool,
+	mhc *clusterv1.MachineHealthCheck,
+	c *clusterv1.Cluster,
+) ([]*corev1.Node, []*clusterv1.Machine, func()) {
+
+	var (
+		nodes    []*corev1.Node
+		machines []*clusterv1.Machine
+	)
+
+	for i := 0; i < n; i++ {
+		machine := newMachine(mhc, c)
+		infraMachine, providerID := newInfraMachine(&machine)
+		node := newNode()
+		if !healthy {
+			setNodeUnhealthy(&node)
+		}
+		g.Expect(testEnv.Create(ctx, infraMachine)).To(Succeed())
+		fmt.Printf("inframachine created: %s\n", infraMachine.GetName())
+		machine.Spec.InfrastructureRef = corev1.ObjectReference{
+			APIVersion: infraMachine.GetAPIVersion(),
+			Kind:       infraMachine.GetKind(),
+			Name:       infraMachine.GetName(),
+		}
+		// TODO: Add comment on what's happening here. I think we are making
+		// the InfraMachine look like its healthy and ready.
+		infraMachinePatch := client.MergeFrom(infraMachine.DeepCopy())
+		g.Expect(unstructured.SetNestedField(infraMachine.Object, true, "status", "ready")).To(Succeed())
+		g.Expect(testEnv.Status().Patch(ctx, infraMachine, infraMachinePatch)).To(Succeed())
+
+		g.Expect(testEnv.Create(ctx, &machine)).To(Succeed())
+		fmt.Printf("machine created: %s\n", machine.GetName())
+
+		// Before moving on we want to ensure that the machine has a valid
+		// status. That is, LastUpdated should not be nil.
+		g.Eventually(func() *metav1.Time {
+			k := client.ObjectKey{
+				Name:      machine.GetName(),
+				Namespace: machine.GetNamespace(),
+			}
+			err := testEnv.Get(ctx, k, &machine)
+			if err != nil {
+				return nil
+			}
+			return machine.Status.LastUpdated
+		}, timeout, time.Second).ShouldNot(BeNil())
+
+		if assignNodeRef {
+			machineStatus := machine.Status
+			node.Spec.ProviderID = providerID
+			nodeStatus := node.Status
+			g.Expect(testEnv.Create(ctx, &node)).To(Succeed())
+			fmt.Printf("node created: %s\n", node.GetName())
+
+			nodePatch := client.MergeFrom(node.DeepCopy())
+			node.Status = nodeStatus
+			g.Expect(testEnv.Status().Patch(ctx, &node, nodePatch)).To(Succeed())
```

Comment:
why are we setting again the node status using a patch?

Reply:
Yeah...I was asking myself this question as well 😄 . I didn't know if there was something interesting happening in the controllers that this needed to be set specifically for the MHC controller.

Reply:
After looking at it some more I noticed that we are using the status client to patch the object. I'm wondering if we cannot create an object with Status fields set and we need the status client specifically to patch them in. This is just a guess.